New unit-tests for ISA (55 implemented) (#530)

pukhovv opened the pull request.

pukhovv committed:
Add test MIPS32_instr: beq

pukhovv committed:
Merge remote-tracking branch 'upstream/master'

pukhovv committed:
New unit-tests for ISA (18 implemented) (#530)

pukhovv committed:
Merge remote-tracking branch 'upstream/master'

pukhovv commented:
Please check movz and movn tests.

pavelkryukov commented on simulator/mips/mips_instr.h:
```diff
@@ -409,10 +409,16 @@ class BaseMIPSInstr
             if ( index == 0)
                 v_src1 = value;
-            else
+            else if ( index == 1)
```
What is the motivation for that check?

pukhovv replied:
Possible errors with index.

pavelkryukov replied:
That is called "defensive programming", as you do not trust your callers provide adequate inputs. That pattern has no advantages, as we guarantee that our inputs are always correct by unit test suites.

pukhovv replied:
OK, I'll remove it.

pavelkryukov commented on simulator/mips/t/unit_test.cpp:
```diff
@@ -243,3 +206,454 @@ TEST_CASE( "MIPS32_instr_disasm: Process_Disasm_Branches")
+    instr.set_shamt( 0);
+    instr.execute();
+    CHECK( instr.get_v_dst() == 1);
+
```
These should be different test cases.

pavelkryukov replied:
@pooh64 You still have one test case covering multiple execution patterns. We have to separate it to different test cases.

pukhovv replied:
sll, sra, srav are not implemented in SPIM TT. Should I modify that case(s), repeated in SPIM TT too?

pavelkryukov replied:
They are: https://github.com/MIPT-ILab/mips-traces/blob/master/tt.core32.s#L518

Their behavior is slightly different for MIPS32 and MIPS64, we need to cover both cases.

pukhovv replied:
Now I'm splitting test cases. All written tests that have different behaviour for MIPS32 and MIPS64 currently are fully covered.

pavelkryukov commented on simulator/mips/t/unit_test.cpp:
```diff
@@ -243,3 +206,454 @@ TEST_CASE( "MIPS32_instr_disasm: Process_Disasm_Branches")
+}
+
+/*
+TEST_CASE( "MIPS32_instr: movz")
```
movz/movn are implemented on the register file level, so we are not able to test them here.

pukhovv replied:
Thanks.

pavelkryukov commented on simulator/mips/t/unit_test.cpp:
```diff
@@ -243,3 +206,454 @@ TEST_CASE( "MIPS32_instr_disasm: Process_Disasm_Branches")
+// 19 done
+
+// Not implemented in SPIM TT
+TEST_CASE ( "MIPS32_instr: sll")
```
Could you please sort instructions in alphabet order?

pukhovv replied:
OK, I have used the order from ISA maps.

pavelkryukov commented on simulator/mips/t/unit_test.cpp:
```diff
@@ -243,3 +206,454 @@ TEST_CASE( "MIPS32_instr_disasm: Process_Disasm_Branches")
+// Not implemented in SPIM TT
+TEST_CASE ( "MIPS32_instr: sll")
+{
+    MIPS32Instr instr( "sll");
```
Please move disassembly checks for implemented instructions to one of the test cases.
With that, we can easily track what is not implemented yet.

pukhovv replied:
Implemented in unit_test.cpp?
Should I write (copy) checks like this
`    CHECK(MIPS32Instr(0x01398820).get_disasm() == "add $s1, $t1, $t9");`
in alphabet order?
One more thing: now build requires module builtins from pyhon-future.

pavelkryukov replied:
> now build requires module builtins from pyhon-future.

Yes, I'm aware. Please run `pip install future`

> Should I write (copy) checks like this in alphabet order?

Please move those checks only if you have implemented the "individual" tests. You would have something like:

```c++
TEST_CASE ("add") {
    CHECK(MIPS32Instr(0x01398820).get_disasm() == "add $s1, $t1, $t9");
    MIPS32Instr instr("add");
// your tests here
}
```

and the long list of instruction without "individual tests" — so we can see which tests have to be added.

pavelkryukov commented on simulator/mips/t/unit_test.cpp:
```diff
@@ -243,3 +206,454 @@ TEST_CASE( "MIPS32_instr_disasm: Process_Disasm_Branches")
+    CHECK( instr.get_v_dst() == 0xfee1dead);
+}
+
+// Not implemented in SPIM TT
```
What do you mean by that comment?

pukhovv replied:
There is no test for this instruction in SPIM TT.

pavelkryukov replied:
OK. Probably that's not so important information to post.

pavelkryukov replied:
My point is that comment should draw attention to some danger or strangeness (you discussed it with Kirill). For that — it is not needed.

pavelkryukov commented on simulator/mips/t/unit_test.cpp:
```diff
@@ -243,3 +206,454 @@ TEST_CASE( "MIPS32_instr_disasm: Process_Disasm_Branches")
+    CHECK( instr.get_new_PC() == instr.get_PC() + 8);
+
+    instr.set_new_PC( instr.get_PC() + 4);
+    instr.set_v_src( -1, 0);
```
You cannot just write "-1" to uint32.
Instead, define a constant in the outer scope:
```c++
static constexpr const uint32 MINUS_ONE32 = all_ones<uint32>();
```

pukhovv replied:
Can I just use 0xffffffff?

pavelkryukov replied:
Yes you can, but IMO this is less readable (how many `f`s are there?)

pavelkryukov committed:
Merge branch 'master' into master

pukhovv committed:
Merge remote-tracking branch 'upsteram/master'

pukhovv committed:
New unit-tests for mips ISA (27 implemented) (#530)

pukhovv commented on simulator/mips/t/unit_test.cpp:
```diff
@@ -236,10 +171,621 @@ TEST_CASE( "MIPS32_instr_disasm: Process_Disasm_Branches")
+    instr.execute();
+    CHECK( instr.get_v_dst() == 0);
+
+/*     Overflow exception isn't implemented
```
add and addi must trap if overflow occurs, but there are none in implementation.

pavelkryukov replied:
Please refer to the issue #130 in comments.
It should be very simple to implement them with these tests.

pavelkryukov commented on simulator/mips/mips_instr.h:
```diff
@@ -189,6 +189,10 @@ class BaseMIPSInstr
                 v_src2 = value;
         }
 
+        void set_new_PC( Addr value) { new_PC = value; }
```
If I understand correctly, this method is needed only for re-use of instruction within a single test case.
If you remove re-uses, it won't be necessary anymore.

pavelkryukov commented on simulator/mips/t/unit_test.cpp:
```diff
@@ -236,10 +171,621 @@ TEST_CASE( "MIPS32_instr_disasm: Process_Disasm_Branches")
+    CHECK( instr.get_v_dst() == 0xfee1dead);
+}
+
+/*      Not implemented on the register file level
```
By the way, I found how to test them without RF.
We implement these instruction via masked register update. If condition is not taken, then "get_mask" return zero and nothing is written to the RF. So, you may check "get_mask" output.

pukhovv replied:
Now tests do not pass on the get_mask(). Mask becomes 0 after each operation (all_one at constr.).

pukhovv replied:
Currently added test implementation that uses a mask.

pavelkryukov commented on simulator/mips/t/unit_test.cpp:
```diff
@@ -236,10 +171,621 @@ TEST_CASE( "MIPS32_instr_disasm: Process_Disasm_Branches")
+    instr.set_v_src( 1, 0);
+    instr.set_v_src( 64, 1);
+    instr.execute();
+    CHECK( instr.get_v_dst() == 1);
```
Ooops, looks like we have a real bug here!
Please comment that test case and open an issue.

https://ci.appveyor.com/project/pavelkryukov/mipt-mips/builds/20445056/job/bggpv4cqhvb7fq7d#L512

pukhovv committed:
Requested fixes in unit-tests (#530)

pukhovv committed:
Merge remote-tracking branch 'upsteram/master'

pavelkryukov commented on simulator/mips/mips_instr.h:
```diff
@@ -190,6 +190,8 @@ class BaseMIPSInstr
                 v_src2 = value;
         }
 
+        void set_shamt(uint16 val) { shamt = val; }
```
Please use `set_v_imm`, as shift amount is set here now

pavelkryukov committed:
Merge branch 'master' into master

pukhovv committed:
Merge remote-tracking branch 'upsteram/master'

pukhovv committed:
New unit-tests for ISA (40 implemented) (#530)
	Fixes in jmp cases.
	Fixed movn/movz case.
	Test for 64bit instr in 32bit ctor.
	References for bug-issue (#709).

pukhovv committed:
Merge branch 'master' of https://github.com/pooh64/mipt-mips

pukhovv commented:
I'll split other cases later.

pukhovv committed:
New unit tests for ISA (55 done) (#530). Splitting

pukhovv committed:
Merge remote-tracking branch 'upstream/master'

pukhovv committed:
Some other tests splitted, #709 tests disabled. (#530)

pavelkryukov committed:
Merge branch 'master' into master

pavelkryukov commented:
Vsevolod, could you please address current review feedback first before submitting any new tests?
Your new test cases replicate some issues of the old test cases. We have not reviewed them yet as we are waiting for ability to refer to test cases unambiguously.

pukhovv commented:
@pavelkryukov I will correct all the issues and fully prepare these 55 tests for the final review and merge. At the moment i'm splitting tests into different cases and searching for any other errors.

pavelkryukov committed:
Merge branch 'master' into master

pukhovv committed:
Merge remote-tracking branch 'upstream/master'

pukhovv committed:
Split tests, fix different misprints (#530)

pukhovv committed:
Merge branch 'master' of https://github.com/pooh64/mipt-mips

pavelkryukov commented on simulator/mips/t/unit_test.cpp:
```diff
@@ -236,10 +146,2252 @@ TEST_CASE( "MIPS32_instr_disasm: Process_Disasm_Branches")
+
+////////////////////////////////////////////////////////////////////////////////
+
+TEST_CASE ( "MIPS32_instr: add test_1")
```
Please use more verbose description, like `MIPS32_Instr: add two zeroes` or `MIPS32_Instr: add overflow`

pukhovv replied:
Done.

pavelkryukov commented on simulator/mips/t/unit_test.cpp:
```diff
@@ -236,10 +146,2252 @@ TEST_CASE( "MIPS32_instr_disasm: Process_Disasm_Branches")
+}
+////////////////////////////////////////////////////////////////////////////////
+
+/*      Not implemented on the register file level
```
Those may be deleted. We test masked writes on RF level instead.

pukhovv replied:
Done.

pavelkryukov commented on simulator/mips/t/unit_test.cpp:
```diff
@@ -236,10 +146,2252 @@ TEST_CASE( "MIPS32_instr_disasm: Process_Disasm_Branches")
+}
+////////////////////////////////////////////////////////////////////////////////
+
+TEST_CASE( "MIPS32_instr: tge test_l")
```
Please expand `test_l` to something more verbose, like `test if less`. You can easily do it with substitution.

pukhovv replied:
Done.

pavelkryukov reviewed (commented):
Please see my notes regarding the code form.
@kkorolev will review test bodies.
Also, please drop a list of instructions which are implemented, but not tested yet on the #530  page for people who are going to continue.

pavelkryukov commented on simulator/mips/t/unit_test.cpp:
```diff
@@ -71,98 +40,44 @@ TEST_CASE( "MIPS32_instr: Divmult")
     CHECK(MIPS32Instr("ddiv").is_divmult());
```
Please move these checks to the per-instruction tests if they exist.
```c++
TEST_CASE( "MIPS32_Instr: mult ... ")
{
// ...
  CHECK( instr.is_divmult());
}
```

pukhovv replied:
Done.

pukhovv committed:
Requested fixes in code form (#530)

pukhovv committed:
Change shift-variable overflow tests (#530)

kkorolev reviewed (approved):
Looks good enough for me

kkorolev committed:
Merge branch 'master' into master
